### PR TITLE
Meta bitfields

### PR DESCRIPTION
**Pull request by vlrpl**
Adds bitfield support for LE ARCH

### CODE REVIEW

**Review by atenart — COMMENTED** [review]
A few high level comments:
- While it's not 100% related, I think allowing operands like `&` would be very nice when working with bitfields, as that would be the main use IMO. We should also find a way to test a bit is not set.
- First patch also includes unrelated changes; those could be split.
- Documentation should be updated too.
- I think some comments could be added here and there to help following the code.

File: src/core/filters/meta/bpf/include/meta_filter.h
Comment (vlrpl) [reply src/core/filters/meta/bpf/include/meta_filter.h:231]: yep, this was changed last minute before pushing the draft to make clang front end happy and I didn't update the branch in the draft (should be the equivalent of if (byo * 8 > bio)).

**Comment by vlrpl** [comment]
> A few high level comments:
> 
>     * While it's not 100% related, I think allowing operands like `&` would be very nice when working with bitfields, as that would be the main use IMO. We should also find a way to test a bit is not set.
> 

I think it's fairly easy to add simple '&' support (for unsigned only I'd say). 
The idea was to tackle this part with #317, but it will not make it for this release (could require parsing rework).
Not sure we might want a separate operator. I'll take a look.
At the moment, I open this PR as draft for 2 reasons: it's a draft (no surprise here :) , we should decide if we want to pick this for the next release.

>     * First patch also includes unrelated changes; those could be split.
> 
>     * Documentation should be updated too.
> 

both the above make sense

>     * I think some comments could be added here and there to help following the code.

will add some notes

**Comment by atenart** [comment]
> > A few high level comments:
> > ```
> > * While it's not 100% related, I think allowing operands like `&` would be very nice when working with bitfields, as that would be the main use IMO. We should also find a way to test a bit is not set.
> > ```
> 
> I think it's fairly easy to add simple '&' support (for unsigned only I'd say). The idea was to tackle this part with #317, but it will not make it for this release (could require parsing rework). Not sure we might want a separate operator. I'll take a look. At the moment, I open this PR as draft for 2 reasons: it's a draft (no surprise here :) , we should decide if we want to pick this for the next release.

If we are planning to add support for `&` soon (ie. next release) I think we can add this for 1.4 as it'll still provide basic bitfield support.

**Comment by vlrpl** [comment]
still just a draft. Will update to a formal version soon

**Comment by vlrpl** [comment]
> > > A few high level comments:
> > > ```
> > > * While it's not 100% related, I think allowing operands like `&` would be very nice when working with bitfields, as that would be the main use IMO. We should also find a way to test a bit is not set.
> > > ```
> > 
> > 
> > I think it's fairly easy to add simple '&' support (for unsigned only I'd say). The idea was to tackle this part with #317, but it will not make it for this release (could require parsing rework). Not sure we might want a separate operator. I'll take a look. At the moment, I open this PR as draft for 2 reasons: it's a draft (no surprise here :) , we should decide if we want to pick this for the next release.
> 
> If we are planning to add support for `&` soon (ie. next release) I think we can add this for 1.4 as it'll still provide basic bitfield support.

yes, the plan is to introduce `&` plus other features targeting 1.5

File: src/core/filters/meta/bpf/include/meta_filter.h
Comment (atenart) [inline src/core/filters/meta/bpf/include/meta_filter.h:198]: I don't get why the `%` above. Shouldn't we just remove the offset in bits?

Nit: would have name the function `extract_bf` for more clarity, but of course not a blocker at all, only personal taste :)

File: src/core/probe/bpf/include/common_defs.h
Comment (atenart) [inline src/core/probe/bpf/include/common_defs.h:113]: I think `m + n` could overflow here; instead `1 + ((m) - 1) / (n)` can be used. Not that it would happen here but if that is reused in other code paths.

File: src/core/filters/meta/bpf/include/meta_filter.h
Comment (atenart) [inline src/core/filters/meta/bpf/include/meta_filter.h:112]: Nit: I think this could be in a dedicated commit with an explanation about what this fixes.

File: src/core/filters/meta/bpf/include/meta_filter.h
Comment (vlrpl) [reply src/core/filters/meta/bpf/include/meta_filter.h:198]: we discussed offline the first part.
The new name sgtm.

File: src/core/filters/meta/bpf/include/meta_filter.h
Comment (vlrpl) [reply src/core/filters/meta/bpf/include/meta_filter.h:112]: ah, right. Forgot to split this.

Comment (vlrpl) [reply src/core/filters/meta/bpf/include/meta_filter.h:112]: looking at this twice, this is a leftover and should be dropped.
It was intended to be a replacement/slightly different approach for the previous `|| !ptr)`  (part of "bpf logger" and written/updated in parallel with this PR). 
This small redundant check fell through the cracks once eBPF logger got merged. Thanks for spotting this, though.

**Review by atenart — APPROVED** [review]
Nice improvement, thanks!

**Pull request merged**
